# Stock Market Analytics Zoomcamp
## Module 1 Homework

In [1]:
!pip install yfinance pandas-datareader plotly matplotlib


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 73.0/73.0 KB 2.6 MB/s eta 0:00:00
  Using cached pandas_datareader-0.10.0-py3-none-any.whl (109 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 15.7/15.7 MB 4.3 MB/s eta 0:00:00m eta 0:00:010:00:01
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 7.5/7.5 MB 3.4 MB/s eta 0:00:00m eta 0:00:010:01:01
  Using cached numpy-1.26.4-cp310-cp310-macosx_11_0_arm64.whl (14.0 MB)
  Using cached frozendict-2.4.2-cp310-cp310-macosx_11_0_arm64.whl (37 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 11.3/11.3 MB 5.2 MB/s eta 0:00:00m eta 0:00:010:0101
  Using cached lxml-5.2.1-cp310-cp310-macosx_10_9_universal2.whl (8.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.0/3.0 MB 5.1 MB/s eta 0:00:00m eta 0:00:010:01:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Using cached appdirs-1.4.4-py2.py3-none-any.whl (9.6 kB)
  Using cached html5lib-1.1-py2.py3-none-any.

In [2]:
import numpy as np
import pandas as pd
import yfinance as yf
import pandas_datareader as pdr
import plotly.graph_objs as go
import plotly.express as px
import time
from datetime import date

# Question 1: [Macro] Average growth of GDP in 2023
**What is the average growth (in %) of GDP in 2023?**

Download the timeseries Real Gross Domestic Product (GDPC1) from FRED (https://fred.stlouisfed.org/series/GDPC1). Calculate year-over-year (YoY) growth rate (that is, divide current value to one 4 quarters ago). Find the average YoY growth in 2023 (average from 4 YoY numbers). Round to 1 digit after the decimal point: e.g. if you get 5.66% growth => you should answer 5.7

In [3]:
end = date.today()
print(f'Year = {end.year}; month= {end.month}; day={end.day}')

start = date(year=end.year-70, month=end.month, day=end.day)
print(f'Period for indexes: {start} to {end} ')

gdpc1_df = pdr.DataReader("GDPC1", "fred", start=start, end=end)


Year = 2024; month= 4; day=21
Period for indexes: 1954-04-21 to 2024-04-21 


In [4]:
gdpc1_df['gdpc1_yoy'] = gdpc1_df.GDPC1/gdpc1_df.GDPC1.shift(4)-1
gdpc1_2023_df = gdpc1_df[gdpc1_df.index.year == 2023]
avg_gdpc1_yoy_2023 = round(gdpc1_2023_df['gdpc1_yoy'].mean()*100, 1)
print('Average YoY growth in 2023:', avg_gdpc1_yoy_2023)

Average YoY growth in 2023: 2.5


# Question 2. [Macro] Inverse "Treasury Yield"
**Find the min value of (dgs10-dgs2) after since year 2000 (2000-01-01) and write it down as an answer, round to 1 digit after the decimal point.**

Download DGS2 and DGS10 interest rates series (https://fred.stlouisfed.org/series/DGS2, https://fred.stlouisfed.org/series/DGS10). Join them together to one dataframe on date (you might need to read about pandas.DataFrame.join()), calculate the difference dgs10-dgs2 daily.

(Additional: think about what does the "inverted yield curve" mean for the market and investors? do you see the same thing in your country/market of interest? Do you think it can be a good predictive feature for the models?)

In [5]:
import pandas as pd


In [6]:
dgs2_df = pdr.DataReader("DGS2", "fred", start=date(year=2000, month=1, day=1))
dgs10_df = pdr.DataReader("DGS10", "fred", start=date(year=2000, month=1, day=1))
print(dgs2_df)
print(dgs10_df)
df_joined = dgs2_df.join(dgs10_df)
df_joined['dgs10-dgs2'] = df_joined['DGS10']-df_joined['DGS2']
print(df_joined)
min_dgs10_dgs2 = round(df_joined['dgs10-dgs2'].min(), 1)
print('Min value of (dgs10-dgs2) after since year 2000 (2000-01-01)', min_dgs10_dgs2)

            DGS2
DATE            
2000-01-03  6.38
2000-01-04  6.30
2000-01-05  6.38
2000-01-06  6.35
2000-01-07  6.31
...          ...
2024-04-12  4.88
2024-04-15  4.93
2024-04-16  4.97
2024-04-17  4.93
2024-04-18  4.98

[6339 rows x 1 columns]
            DGS10
DATE             
2000-01-03   6.58
2000-01-04   6.49
2000-01-05   6.62
2000-01-06   6.57
2000-01-07   6.52
...           ...
2024-04-12   4.50
2024-04-15   4.63
2024-04-16   4.67
2024-04-17   4.59
2024-04-18   4.64

[6339 rows x 1 columns]
            DGS2  DGS10  dgs10-dgs2
DATE                               
2000-01-03  6.38   6.58        0.20
2000-01-04  6.30   6.49        0.19
2000-01-05  6.38   6.62        0.24
2000-01-06  6.35   6.57        0.22
2000-01-07  6.31   6.52        0.21
...          ...    ...         ...
2024-04-12  4.88   4.50       -0.38
2024-04-15  4.93   4.63       -0.30
2024-04-16  4.97   4.67       -0.30
2024-04-17  4.93   4.59       -0.34
2024-04-18  4.98   4.64       -0.34

[6339 rows x 3 columns]
Mi

# Question 3. [Index] Which Index is better recently?

**Compare S&P 500 and IPC Mexico indexes by the 5 year growth and write down the largest value as an answer (%)**

Download on Yahoo Finance two daily index prices for S&P 500 (^GSPC, https://finance.yahoo.com/quote/%5EGSPC/) and IPC Mexico (^MXX, https://finance.yahoo.com/quote/%5EMXX/). Compare 5Y growth for both (between 2019-04-09 and 2024-04-09). Select the higher growing index and write down the growth in % (closest integer %). E.g. if ratio end/start was 2.0925 (or growth of 109.25%), you need to write down 109 as your answer.

(Additional: think of other indexes and try to download stats and compare the growth? Do create 10Y and 20Y growth stats. What is an average yearly growth rate (CAGR) for each of the indexes you select?)


In [7]:
import yfinance as yf

In [8]:
sp500_df = yf.download(tickers = "^GSPC", start='2019-04-09', end='2024-04-09', interval = "1d")
ipcmex_df = yf.download(tickers = "^MXX", start='2019-04-09', end='2024-04-09', interval = "1d")
# print(sp500_df.tail())
# print(sp500_df.head())


[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [9]:

# Calculate YoY growth for S&P 500 (^GSPC)
sp500_df['yoy_growth'] = (sp500_df['Close']/sp500_df['Close'].shift(252)-1)
sp500_growth = int(sp500_df['yoy_growth'].sum()) # 125.00 - 125%

# sp500_start_date = sp500_df.index[0]
# sp500_end_date = sp500_df.index[-1]
# sp500_start = sp500_df['Adj Close'].iloc[0]
# sp500_end = sp500_df['Adj Close'].iloc[-1]

# ratio [=(max-min)/max]
# sp500_growth = ((sp500_end / sp500_start) - 1) * 100 # 80.75 - 81%
# sp500_growth = ((sp500_end - sp500_start) / sp500_start)*100 # 80.75 - 81%

# print(f"{sp500_start_date.strftime('%B %d, %Y')} - {sp500_start:.2f}")
# print(f"{sp500_end_date.strftime('%B %d, %Y')} - {sp500_end:.2f}")
# print(f"{sp500_growth:.2f}")

# Calculate YoY growth for IPC Mexico (^MXX)
ipcmex_df['yoy_growth'] = (ipcmex_df['Close']/ipcmex_df['Close'].shift(252)-1)
ipcmex_growth = int(ipcmex_df['yoy_growth'].sum())

# ipcmex_start_date = ipcmex_df.index[0]
# ipcmex_end_date = ipcmex_df.index[-1]
# ipcmex_start = ipcmex_df['Adj Close'].iloc[0]
# ipcmex_end = ipcmex_df['Adj Close'].iloc[-1]

# ipcmex_growth = ((ipcmex_end / ipcmex_start) - 1) * 100
# ipcmex_growth = ((ipcmex_end - ipcmex_start) / ipcmex_start)*100

# print(f"{ipcmex_start_date.strftime('%B %d, %Y')} - {ipcmex_start:.2f}")
# print(f"{ipcmex_end_date.strftime('%B %d, %Y')} - {ipcmex_end:.2f}")
# print(f"{ipcmex_growth:.2f}")

if sp500_growth > ipcmex_growth:
    higher_index = 'S&P 500 (^GSPC)'
    growth_percentage = int(round(sp500_growth))
else:
    higher_index = 'IPC Mexico (^MXX)'
    growth_percentage = int(round(ipcmex_growth))

print(f"The index with higher growth is {higher_index} with a growth of {growth_percentage}%.")


The index with higher growth is S&P 500 (^GSPC) with a growth of 125%.


# Question 4. [Stocks OHLCV] 52-weeks range ratio (2023) for the selected stocks
**Find the largest range ratio [=(max-min)/max] of Adj.Close prices in 2023**

Download the 2023 daily OHLCV data on Yahoo Finance for top6 stocks on earnings (https://companiesmarketcap.com/most-profitable-companies/): 2222.SR,BRK-B, AAPL, MSFT, GOOG, JPM.

Here is the example data you should see in Pandas for "2222.SR": https://finance.yahoo.com/quote/2222.SR/history

Calculate maximum-minimim "Adj.Close" price for each stock and divide it by the maximum "Adj.Close" value. Round the result to two decimal places (e.g. 0.1575 will be 0.16)

(Additional: why this may be important for your research?)

In [10]:
_2222sr_df = yf.download(tickers = "2222.SR", start='2023-01-01', end='2023-12-31', interval = "1d")
brkb_df = yf.download(tickers = "BRK-B", start='2023-01-01', end='2023-12-31', interval = "1d")
aapl_df = yf.download(tickers = "AAPL", start='2023-01-01', end='2023-12-31', interval = "1d")
msft_df = yf.download(tickers = "MSFT", start='2023-01-01', end='2023-12-31', interval = "1d")
goog_df = yf.download(tickers = "GOOG", start='2023-01-01', end='2023-12-31', interval = "1d")
jpm_df = yf.download(tickers = "JPM", start='2023-01-01', end='2023-12-31', interval = "1d")

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


In [11]:
_2222sr_rr = round((_2222sr_df['Adj Close'].max()-_2222sr_df['Adj Close'].min())/_2222sr_df['Adj Close'].max(),2)
brkb_rr = round((brkb_df['Adj Close'].max()-brkb_df['Adj Close'].min())/brkb_df['Adj Close'].max(),2)
msft_rr = round((msft_df['Adj Close'].max()-msft_df['Adj Close'].min())/msft_df['Adj Close'].max(),2)
goog_rr = round((goog_df['Adj Close'].max()-goog_df['Adj Close'].min())/goog_df['Adj Close'].max(),2)
jpm_rr = round((jpm_df['Adj Close'].max()-jpm_df['Adj Close'].min())/jpm_df['Adj Close'].max(),2)

print('Largest range ratio of Adj.Close prices in 2023: ', max(_2222sr_rr,brkb_rr,msft_rr,goog_rr,jpm_rr))

Largest range ratio of Adj.Close prices in 2023:  0.42


# Question 5. [Stocks] Dividend Yield
**Find the largest dividend yield for the same set of stocks**

Use the same list of companies (2222.SR,BRK-B, AAPL, MSFT, GOOG, JPM) and download all dividends paid in 2023. You can use get_actions() method or .dividends field in yfinance library (https://github.com/ranaroussi/yfinance?tab=readme-ov-file#quick-start)

Sum up all dividends paid in 2023 per company and divide each value by the closing price (Adj.Close) at the last trading day of the year.

Find the maximm value in % and round to 1 digit after the decimal point. (E.g., if you obtained $1.25 dividends paid and the end year stock price is $100, the dividend yield is 1.25% -- and your answer should be equal to 1.3)

In [12]:
def get_year_dividends_yeld(ticker, year):
  ohlcv_df = yf.download(tickers = ticker, start=f'{year}-12-01', end=f'{year}-12-31', interval = "1d")
  last_adj_close = ohlcv_df['Adj Close'].iloc[-1]
  yf_ticker = yf.Ticker(ticker)
  actions_df = yf_ticker.get_actions()
  dividends_sum = actions_df[actions_df.index.year == year]['Dividends'].sum()
  return round((dividends_sum / last_adj_close)*100, 1)


In [13]:
_2222sr_divs_yeld = get_year_dividends_yeld('2222.SR', 2023)
brkb_divs_yeld = get_year_dividends_yeld('BRK-B', 2023)
aapl_divs_yeld = get_year_dividends_yeld('AAPL', 2023)
msft_divs_yeld = get_year_dividends_yeld('MSFT', 2023)
goog_divs_yeld = get_year_dividends_yeld('GOOG', 2023)
jpm_divs_yeld = get_year_dividends_yeld('JPM', 2023)

div_yelds_2023 = [_2222sr_divs_yeld, brkb_divs_yeld, aapl_divs_yeld, msft_divs_yeld, goog_divs_yeld, jpm_divs_yeld]

print('Largest 2023 dividends yield: ', max(div_yelds_2023), '%')

[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed
[*********************100%%**********************]  1 of 1 completed


Largest 2023 dividends yield:  2.8 %


# Question 6. [Exploratory] Investigate new metrics
**Free text answer**

Download and explore a few additional metrics or time series that might be valuable for your project and write down why (briefly).



# Question 7. [Exploratory] Time-driven strategy description around earnings releases
**Free text answer**

Explore earning dates for the whole month of April - e.g. using YahooFinance earnings calendar (https://finance.yahoo.com/calendar/earnings?from=2024-04-21&to=2024-04-27&day=2024-04-23). Compare with the previous closed earnings (e.g., recent dates with full data https://finance.yahoo.com/calendar/earnings?from=2024-04-07&to=2024-04-13&day=2024-04-08).

Describe an analytical strategy/idea (you're not required to implement it) to select a subset companies of interest based on the future events data.